Bugzilla integration

KaratekHD opened the pull request:
When done, this should close #2. For know, it does not yet support searching bugzilla but requires you to know a bug ID.

KaratekHD commented:
With these changes the tests should work in GitHub actions too.

why-not-try-calmer reviewed (changes requested):
Added the missing function return types and now just a few general comments on `bugs.py`:
- every parsing, casting or network request should be protected by `try...except... (finally)` guards, including line 45 
- `get_bug_from_bugzilla` should be made async (for example using `helpers.sync_utils.as_async`) 
- I suggest not caching anything right now. The caching strategies will probably need a more complex treatment than available (#29 kicks off a proposal for that). And this messes up with your design
- I think the parsing is good but `get_bug_from_bugzilla` and `universal_search` would be more readabl if you started from the other direction, i.e. first define a dataclass with mandatory vs optional attributes after having looked up the API specs, and assign them separately. If there is no optional field, it's even better! Instead of a simple dataclass you can sublass NamedTuple or even use built-in nametuples since the data is immutable anyway.

Also a few details:
- not used: `import asyncio` (25), `term` (108)

KaratekHD commented:
> Added the missing function return types and now just a few general comments on `bugs.py`:
> 
>     * every parsing, casting or network request should be protected by `try...except... (finally)` guards, including line 45
> 
>     * `get_bug_from_bugzilla` should be made async (for example using `helpers.sync_utils.as_async`)
> 
>     * I suggest not caching anything right now. The caching strategies will probably need a more complex treatment than available ([Added cache middleware + services manager. #29](https://github.com/KaratekHD/defrag/pull/29) kicks off a proposal for that). And this messes up with your design
> 
>     * I think the parsing is good but `get_bug_from_bugzilla` and `universal_search` would be more readabl if you started from the other direction, i.e. first define a dataclass with mandatory vs optional attributes after having looked up the API specs, and assign them separately. If there is no optional field, it's even better! Instead of a simple dataclass you can sublass NamedTuple or even use built-in nametuples since the data is immutable anyway.
> 
> 
> Also a few details:
> 
>     * not used: `import asyncio` (25), `term` (108)

Alright, will work the other changes in later

KaratekHD commented:
> I think the parsing is good but `get_bug_from_bugzilla` and `universal_search` would be more readabl if you started from the other direction, i.e. first define a dataclass with mandatory vs optional attributes after having looked up the API specs, and assign them separately. If there is no optional field, it's even better! Instead of a simple dataclass you can sublass NamedTuple or even use built-in nametuples since the data is immutable anyway.

Sorry, I do not quite understand what exactly you want me to change, could you maybe propose a change? There is a button for that while reviewing.

why-not-try-calmer commented on defrag/modules/bugs.py:
```diff
@@ -0,0 +1,184 @@
+        "module": "Bugzilla",
+        "description": "Get information about bugs on bugzilla.opensuse.org"}
+
+
```
The suggestion to clarify your implementation was simply to go for something like this:
```python
@dataclass
class BugZillaEntry:
    id: int
    product: Optional[str]
    # etc, etc.
```
Since the data is immutable anyway you can even use a plain `Namedtuple` instead of a dataclass. Or you can sublcass `Namedtuple` if you feel like you need to override or add behaviour.

KaratekHD replied:
Done.

why-not-try-calmer commented on defrag/modules/bugs.py:
```diff
@@ -0,0 +1,184 @@
+
+
+@app.get("/" + __MOD_NAME__ + "/bug/")
+async def get_bug(id: int, response: Response) -> Dict[str, str]:
```
then you are able to be more specific in the return type:
```python
async def get_bug(id: int, response: Response) -> Optional[BugZillaEntry]:
```
if you feel like you need to invalidate the whole thing when unable to assign certain fields/attributes (but then should they be mandatory? >:D ) or just 
```python
async def get_bug(id: int, response: Response) -> BugZillaEntry:
```
if you feel you will always be able to build an entire object. This will make the implementation more readable hence easier to maintain.

KaratekHD replied:
Done

why-not-try-calmer commented on defrag/modules/helpers/exceptions.py:
```diff
@@ -0,0 +1,35 @@
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+class DefragException(Exception):
```
I would have preferred a distinct PR for this but I agree it's a good idea.

KaratekHD replied:
Sorry

why-not-try-calmer reviewed (changes requested):
A couple of suggestions:
- rebase onto `dev`. Then you'll have access to request.py, so that you can rewrite line 137 and below with it.
- perhaps move "all endpoints call" (@app.get) together at the end of the file so that we know at one glance which routes the module provides?
- 94-103: can test simply with `if s:`
- your comment on 169: we don't have to use decorators at all for caching, but this relates to my pending PR so ok we'll talk about this
- I think you can make the `Dict[...,...]` return types more clear if you use either a dataclass or a nametuple (nice because immutable and super light) or subclass `nametuple`

Apart from this it's OK for me so if you make these changes you can merge as far as I am concerned. Good job!

KaratekHD commented:
> A couple of suggestions:
> 
>     * rebase onto `dev`. Then you'll have access to request.py, so that you can rewrite line 137 and below with it.
> 
>     * perhaps move "all endpoints call" (@app.get) together at the end of the file so that we know at one glance which routes the module provides?
> 
>     * 94-103: can test simply with `if s:`
> 
>     * your comment on 169: we don't have to use decorators at all for caching, but this relates to my pending PR so ok we'll talk about this
> 
>     * I think you can make the `Dict[...,...]` return types more clear if you use either a dataclass or a nametuple (nice because immutable and super light) or subclass `nametuple`
> 
> 
> Apart from this it's OK for me so if you make these changes you can merge as far as I am concerned. Good job!

Note to myself: http://sirius.scandic.priv/issues/29

KaratekHD commented:
>     * 94-103: can test simply with `if s:`

I tried that, but we can not do that, as it always is initialized, just empty if nothing is in there

why-not-try-calmer commented:
> > ```
> > * 94-103: can test simply with `if s:`
> > ```
> 
> I tried that, but we can not do that, as it always is initialized, just empty if nothing is in there

Really? https://replit.com/join/suwvvlwvsm-nycticorax

why-not-try-calmer committed:
Reset because too many commits

why-not-try-calmer committed:
Adrien/jens/proposal/bugzilla (#43)

* init

.

some comments

- simplified and improved cache
- WIP, be done in a few hours

to rebase

.

.

added documentation and comments

some more comments

* - fixed some typos in my code

* - bumped requirements

KaratekHD committed:
Fix logging

why-not-try-calmer committed:
all tests pass

why-not-try-calmer committed:
updated requirements
